### ISSUE TITLE
ENH: Enhanced support for scipy.interpolate.RbfInterpolator norms

### ISSUE THREAD

**Issue by smeingast** [scipy/scipy#17044]
### Is your feature request related to a problem? Please describe.

This request is not related to a problem

### Describe the solution you'd like.

It is already possible to supply a specific norm to the current implementation of Rbf. The norm can be a user-defined callable, however, this is very slow and the recommendation is to use any of the metrics that are supplied with `scipy.spatial.distance.cdist`. 

Some of the metrics in cdist support keyword args, such as the Minkowski metric that can take a weight or the standardised Euclidean metric that supports a variance vector. However, the current implementation of Rbf does not allow to pass any keyword arguments to the metrics defined in cdist.

### Describe alternatives you've considered.

An alternative solution would be to allow a precomputed distance and pass a distance matrix, similar to some implementations in scikit-learn. See, for instance the Nearest Neighbour implementation: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html

### Additional context (e.g. screenshots, GIFs)

See metrics and kwargs for cdist:
(see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html)

**Comment by tupui** [comment]
Thank you for reporting @smeingast. As noted in the doc, `RBFInterpolator` should be used in new code instead. Hence I am not sure if we would want to do anything with `Rbf`.

**Comment by tupui** [comment]
Correct, the kernel is being called with `kernel_func(np.linalg.norm(x[i] - y[i]))`. So maybe we could add a metric option. The issue being that this part is handled in Pythran at the moment.

**Comment by ev-br** [comment]
Overall, this sounds reasonable. A big question is, like Pamphile said, how to do it in a clean way with pythran implementations. A pull request is most welcome @smeingast !

> Hence I am not sure if we would want to do anything with Rbf. (I am wondering if we should instead think about starting a deprecation cycle.)

That, or declaring it _legacy_ in the sense of https://mail.python.org/archives/list/scipy-dev@python.org/thread/Z33JO27EX6BX7LVBI6CW6MX64VQZIJ66/#6GLVRHX5XBTZNL62LFN4YPGACY5BDPLA
Would be nice to start on the infra for this status BTW, as a ReST directive etc. Is it something you'd be interested @tupui?

**Comment by tupui** [comment]
Sorry for the delay @ev-br. Sure I can help. I did not reply to the emails but I am +1 on either deprecating (which I always prefer) or marking it legacy. From the emails, seems like legacy is preferred? Can you confirm? And maybe we should put a warning but just for 1 or 2 releases and then remove the warning? That would give some incentive to users to migrate. Lastly, I am also on the side of not doing anything anymore for legacy stuff. When it breaks for good, then we remove it, simple as that.

**Comment by ev-br** [comment]
Rbf specifically we can deprecate I think if there's a general preference. While I'm not an rbf user, I suspect that rnfinterpolator offers a strict superset, fixes long-standing Rbf problems and migration is not too onerous. 
Maybe send an email -dev list to gauge the mood? 

For the legacy designation, like I was saying elsewhere, there are other candidates (interp1d, the UnivariateSpline family maybe more). Once the infra is there (no warnings, just a  way to make a big fat note in the docs, I'd say) , it will be used for sure

**Comment by dopplershift** [comment]
We have some code over in MetPy that uses `interpolate.Rbf`. I'm happy to migrate it, but would love it if this didn't start spitting out deprecation warnings until I had a full 2-year (NEP-29 style) window of Scipy releases that had the new code. (`RBFinterpolator` was added in 1.7.0, released in June 2021.)

**Comment by tupui** [comment]
Thank you for the feedback @dopplershift. Normally we would produce warning if we deprecate something. Our usual workflow is, deprecate in doc + code with a warning in version n, remove in n+2 as per NEP29. We rarely deprecate first in the doc, then wait n+2 and then warning and n+2 (most users not reading docs, we cannot avoid n+2 with a warning). That would mean n+4 which is 2 years...

**Comment by dopplershift** [comment]
Right, so I guess what I'm asking is don't deprecate yet. I understand if that's not what you all want to do. I'm just trying to avoid the part where I'm staring at deprecation warnings I can't do anything about for 6+ months, unless I want to force my users to upgrade their scipy versions. If I have to stare at warnings (and silently grumble about y'all 😉 ) then that's what I'll do.

Just my $0.02, do with it as you want.

**Comment by tupui** [comment]
Why would you not filter out the warning? What we are going to do is pretty standard (not just in SciPy) and we are deprecating things like that all the time.

**Comment by dopplershift** [comment]
> Why would you not filter out the warning? What we are going to do is pretty standard (not just in SciPy) and we are deprecating things like that all the time.

It's not nearly as often that things are deprecated in favor of something that was recently added. (That's not to say y'all are doing something weird, I just don't have to deal with it regularly.) We usually try to address deprecation warnings as soon as they're encountered. I don't love the warning filter because that becomes something else that needs to be maintained. It's not like our test suite is currently warning free.

Don't consider this an impassioned plea. Just offering my feedback and thoughts since you asked on the mailing list.

**Comment by ev-br** [comment]
FWIW, I'd think it's reasonable ask to postpone the deprecation until the next release cycle, so that Rbf starts emitting deprecation warnings around summer 2023. Would you please respond to the mailing list thread @dopplershift , so that the discussion is not scattered?

**Comment by EricPutney** [comment]
Hi all, +1 to this feature request for RBFInterpolator. It would be incredibly useful to be able to specify a custom norm again. For my use case, just a subset of inputs live in euclidean space but I still need to identify nearest neighbors (in that space) for memory reasons.

**Comment by ev-br** [comment]
cc https://github.com/scipy/scipy/issues/17044#issuecomment-1266404715

This is still a very welcome enhancement, where the technical difficulty is how to integrate with `pythran`. A pull request is definitely very welcome!
Maybe @serge-sans-paille could weigh in on how to integrate custom callables with pythran?

**Comment by serge-sans-paille** [comment]
@ev-br would https://pythran.readthedocs.io/en/latest/MANUAL.html#capsule-corp do the trick? There's an example usage here https://pythran.readthedocs.io/en/latest/examples/Third%20Party%20Libraries.html#With-Pythran and in this blog post: https://serge-sans-paille.github.io/pythran-stories/the-capsule-corporation.html#using-pythran-to-generate-a-capsule

**Comment by ev-br** [comment]
Possibly, yes. These links seem to be about how to generate a PyCapsule with pythran, mostly?
For this feature request and https://github.com/scipy/scipy/issues/17294, what's still not clear at least to me is how to provide a callable to a pythran-compiled function. 

The target API seems to be that the `kernel` argument of https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.RBFInterpolator.html
is now a string, and the user requests are --- AFAIU --- to be able to overload with a user-supplied callable. Maybe we could restrict it to be a pythran-generated LowLevelCallable if that simplifies something.

**Comment by serge-sans-paille** [comment]
> Maybe we could restrict it to be a pythran-generated LowLevelCallable if that simplifies something.

There is a specific pythran syntax to state "this argument is a function pointer in a capsule"

```
#pythran export brief(int, int((int, str), str set)):
def brief(val, capsule):
    return capsule((val, "doctor"), {"some"})
```

the capsule need not being generated by pythran, but it cannot use any python function (i.e. it must contain only native code, no call to the Python C API).
